library/swift/Engine.swift

```diff
@@ -9,4 +9,10 @@ public protocol Engine: AnyObject {
 
   /// A client for recording time series metrics.
   func pulseClient() -> PulseClient
+
+  ///
+  /// Flush the stats sinks outside of a flushing interval.
+  /// Note: stats flushing may not be synchronous.
```

Comment:
this is just out of curiosity, since i read "may not be synchronous" a few times in the docs: when is the stats flushing synchronous and when is it async?

Reply:
The way it works is that if a flushStats is issued before the server has initialized it will flush directly, otherwise it will schedule a histogram merge on all the worker threads and do the flush once this has completed. That said, since we issue this flush request by putting it on the EM dispatcher, this top level function will never block even if we do the synchronous flush code path.

Maybe I should update this comment (I took it from the revert) to be clearer here?

Reply:
👍

just a follow up clarifying question so im clear on it:
> if a flushStats is issued before the server has initialized it will flush directly

i might be wrong, but im assuming there won't be "real" stats flush before the server has initialized?

Reply:
So the Envoy handler has logic for doing a flush before the server has been initialized: https://github.com/envoyproxy/envoy/blob/main/source/server/server.cc#L208-L229

This makes it possible to do a stat merge that doesn't rely on merging histograms (since worker threads may not have been created yet).

If a merge comes in before `server_` has been set on the EM engine it will be a noop since we never dispatch a flushStats at all

So there are three results: 1) before EM/envoy's startup process has started (noop) 2) during EM/envoy init phase (direct, no histogram merge) and 3) after EM/envoy init (histogram merge, async flush)

Reply:
👌🙏